Management command to keep software plan consistent

jingcheng16 opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
Write a management command, to pull the list of every customer billing account and for each account, what software plan version each domain is on. After Ops confirmed the list, run the command again with "--update" will update all inconsistent software plan to the main billing domain's software plan.



## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
Ticket: https://dimagi-dev.atlassian.net/browse/SAAS-14842

Some context:
- For a customer billing account (Enterprise account), the main billing domain should have `do_not_invoice=False`, and all other domain should have `do_not_invoice=True`. This will guarantee we only charge plan cost once for each customer billing account.
- We don't always want to automatically upgrade for all billing account, so a prompt is provided for flexiblity

Upgrade Logic confirmed by Ops:
- Identify the main billing domain and the main billing subscription
- Scan through all subscription, that are 1) still active 2) type PRODUCT or IMPLEMENTATION, if the software plan is different from the main billing subscription’s software plan, update it (if the domains share the same account)

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
- Tested locally
- Write test for this command
- Ops will always review the output first before we run the command with `--update`

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

Adding new test `TestKeepSoftwarePlanConsistentManagementCommand` to basically make sure the command works.

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
No QA

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

jingcheng16 committed:
management command

jingcheng16 committed:
Add need update column

jingcheng16 committed:
Prompt for billing account name

jingcheng16 committed:
Function to upgrade single subscription

jingcheng16 committed:
Add test to make sure only type product or implementation get upgraded

jingcheng16 committed:
Better assertion error message

jingcheng16 committed:
Update software plan

jingcheng16 committed:
Test management command

jingcheng16 committed:
Remove subscription type limit on upgrading software plan

jingcheng16 committed:
Add comment

millerdev commented on corehq/apps/accounting/management/commands/list_customer_billing_account_software_plan.py:
```diff
@@ -0,0 +1,98 @@
+                break
+            else:
+                try:
+                    BillingAccount.objects.get(name=account_name, is_active=True, is_customer_billing_account=True)
```
What do you think of saving the value returned here rather than doing another query at the end to lookup all entered names?

jingcheng16 replied:
Yes, this will improve the performance. Fixed here: fadda1231b19693fefad5af269f5a7a05035680f

millerdev commented on corehq/apps/accounting/management/commands/list_customer_billing_account_software_plan.py:
```diff
@@ -0,0 +1,98 @@
+                    billing_account_names.append(account_name)
+
+        if 'ALL' in billing_account_names:
+            customer_billing_account = BillingAccount.objects.filter(
+                is_customer_billing_account=True, is_active=True)
+        else:
+            customer_billing_account = BillingAccount.objects.filter(
```
What do you think of changing the name `customer_billing_account` to be plural since it is a collection of accounts?

jingcheng16 replied:
Thank you! My oversight. Fixed here: 847178f4744ee3968b1b0a76a097ec4bd116608c

millerdev commented on corehq/apps/accounting/management/commands/list_customer_billing_account_software_plan.py:
```diff
@@ -0,0 +1,98 @@
+                        main_subscription = None
+                        csv_main_subscription = "Multiple Main Subscriptions"
+                    for subscription in subscriptions:
+                        need_update = False
+                        update_status = ''
+                        if (
+                            main_subscription
+                            and subscription.plan_version != main_subscription.plan_version
+                            and subscription.service_type in (SubscriptionType.PRODUCT,
+                                                              SubscriptionType.IMPLEMENTATION)
+                        ):
+                            need_update = True
```
nit
```suggestion
                        update_status = ''
                        need_update = bool(
                            main_subscription
                            and subscription.plan_version != main_subscription.plan_version
                            and subscription.service_type in (SubscriptionType.PRODUCT,
                                                              SubscriptionType.IMPLEMENTATION)
                        )
```

millerdev reviewed (commented):
> Ops will always review the output first before we run the command with `--update`

Imagine a scenario where some new account is created or somehow gets into a state that it would be updated after the command is run the first time without `--update`, so it was not on the original list reviewed by ops. It seems like we should not update that account until it is reviewed. One idea that comes to mind is to change the `--update` argument to accept a file containing a list of accounts, and only accounts in that file would be updated.

biyeun commented on corehq/apps/accounting/management/commands/list_customer_billing_account_software_plan.py:
```diff
@@ -0,0 +1,98 @@
+                    # Value for csv column Main Subscription
+                    csv_main_subscription = ""
+                    try:
+                        main_subscription = subscriptions.filter(do_not_invoice=False).get()
```
what happens in the case that this returns more than one subscription? There are some `BillingAccounts` that have this situation

biyeun replied:
nevermind, I see the following except

biyeun commented on corehq/apps/accounting/management/commands/list_customer_billing_account_software_plan.py:
```diff
@@ -0,0 +1,98 @@
+                                                              SubscriptionType.IMPLEMENTATION)
+                        ):
+                            need_update = True
+                        if update and need_update:
```
I like Daniel's suggestion of passing in a list or file of billing account ids/names that can be updated. Maybe it might also help with the phrasing of this condition. instead of just `update` maybe `allow_update` might be more clear? nit on `need_update` to `needs_update`

jingcheng16 replied:
Thank you Biyeun! I definitely need to pay more attention to naming. Fixed here: d8f759b977f419e58772259cc5159a5f80007412

jingcheng16 committed:
Refactor: Simplify need_update assignment using explicit bool conversion

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

jingcheng16 committed:
Refactor: rename customer_billing_account to be plural since it is a collection of accounts

jingcheng16 committed:
Refactor: Optimize billing account retrieval process

Previously, we queried the database twice for billing accounts: once to verify existence and once to retrieve the actual accounts. This refactor consolidates these operations, immediately saving the result of the successful query to an array.

jingcheng16 committed:
Refactor: rename update to allow_update, need_update to needs_update

millerdev commented on corehq/apps/accounting/management/commands/list_customer_billing_account_software_plan.py:
```diff
@@ -0,0 +1,94 @@
+                                 "type 'ALL' for all customer billing account, "
+                                 "type 'DONE' when finished:\n")
+            if account_name.upper() == 'ALL':
+                customer_billing_accounts = ['ALL']
+                break
+            elif account_name.upper() == 'DONE':
+                break
+            else:
+                try:
+                    account = BillingAccount.objects.get(name=account_name,
+                                                         is_active=True, is_customer_billing_account=True)
+                except BillingAccount.DoesNotExist:
+                    print(f"Customer billing account {account_name} does not exist")
+                else:
+                    customer_billing_accounts.append(account)
+
+        if 'ALL' in customer_billing_accounts:
+            customer_billing_accounts = BillingAccount.objects.filter(
+                is_customer_billing_account=True, is_active=True)
```
Looks like `customer_billing_accounts` can contain two different types of objects (string or account). This is a common cause of bugs. Luckily I think this one can be simplified by moving the condition below into this `if` block.
```suggestion
                customer_billing_accounts = BillingAccount.objects.filter(
                    is_customer_billing_account=True, is_active=True)
                break
            elif account_name.upper() == 'DONE':
                break
            else:
                try:
                    account = BillingAccount.objects.get(name=account_name,
                                                         is_active=True, is_customer_billing_account=True)
                except BillingAccount.DoesNotExist:
                    print(f"Customer billing account {account_name} does not exist")
                else:
                    customer_billing_accounts.append(account)
```

jingcheng16 committed:
Refactor: query all the billing accounts when prompt get "ALL"

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

millerdev commented:
> when we update, we will still only update the accounts that confirmed by ops.
> Then, why would we want to change the `--update` argument to accept a file containing a list of accounts?

I was assuming (apparently incorrectly) that a list would first be generated with `ALL`, and then the same `ALL` entry would be used with `--update` when running again after ops' review. It sounds like we will instead send a list to ops, and then run the command with `--update` and manually type each account name that is confirmed by ops. If that's correct, and not too tedious for the person running the script, then I don't have a problem with that process.

> Is the goal here is to force the user always provide a file, when use `--update`?

Yes, that was my idea. Do you think it would be reasonable to disable the `ALL` option when running with `--update`? Otherwise someone may be tempted to use that option and accidentally process accounts that were not approved by ops.

jingcheng16 committed:
Disable the ALL option when running with --update

jingcheng16 commented:
@millerdev 
Hi Daniel, thank you for your suggestion! Disabled the `ALL` option when running with `--update` here: 2a5d35f1a846bed7418e20fe14b8841c1dfb334d

jingcheng16 committed:
Fix grammar issue

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

jingcheng16 commented on corehq/apps/accounting/management/commands/list_customer_billing_account_software_plan.py:
```diff
@@ -0,0 +1,95 @@
+                                 "type 'DONE' when finished:\n")
+            if account_name.upper() == 'ALL':
+                if allow_update:
+                    print("Please specify the exact list of account when running with --update"
```
:joy: My next goal is to write code without grammar mistakes :sweat_smile:. Thank you and fixed in 3c5b1f4

jingcheng16 committed:
Merge branch 'master' into jc/inconsistent-software-plan

jingcheng16 committed:
Fix test error since we now disallow ALL when update is true
